File: day7p1.py
# Find all of the directories with a total size of at most 100000,
# then calculate the sum of their total sizes.
# In the example above,
# these directories are a and e;
# the sum of their total sizes is 95437 (94853 + 584)

def process(command):
    # if isdigit(x[0]):
    sum = 0
    # for d in directories:
    #     if size[d] <= 100000:
    #         sum += size[d]
    return sum

file = open('./day7input.txt', "r")
lines = file.read().splitlines()

directories = {
    '/': []
}
contents = []
current_directory = ['/']

for l in lines:
    x = l.split(" ")
    if x[0] == "$":
        if x[1] == "cd":
            if x[2] == "..":
                current_directory.pop()
            else:
                current_directory.append(x[2])
        if x[1] == "ls":
            cd = "".join(current_directory)
            directories[cd] = directories[cd] + contents
            contents = []

    else:
        contents.append(l)



# Keep track of what directory you start from
# and when they change directory.
    # can add onto the directory tree
    # can pop off the directory.

# Can recognize a file as something that starts with a number.
# Can recognize a directory when it starts with a 'dir'.


# Write a function to traverse the directory structure
# sum sizes.

File: day7p1.js
const fs = require('fs');
const input = fs.readFileSync('./day7input.txt', 'utf8').split('\n');
let currentDirectory = [];
let cache = {};

// Recursively calculate size of directory.
// For each element of our directory,
// we sum file sizes when we encounter a number.
// When we see a directory, we push it to our currentDirectory and recurse.
function sizeOfDirectory(dir, size=0) {
    for(let i=0; i<dir.length; i++) {
        const entry = dir[i];
        if(/^[0-9]*$/.test(entry)) {
            size += parseInt(entry);
        } else if (/^[A-Za-z]*$/.test(entry)) {
            currentDirectory.push(entry);
            size += sizeOfDirectory(cache[currentDirectory.join('/')]);
        }
    }
    currentDirectory.pop();
    return parseInt(size);
}

function command(line) {
    switch(line[1]) {
        case 'cd':
            if(line[2] == '..') {
                currentDirectory.pop();
            } else {
                currentDirectory.push(line[2]);
            }
            break;
        case 'ls':
            cache[currentDirectory.join('/')] = [];
            break;
        default:
            console.log('invalid command');
    }
}

// Loop over input and execute commands
// Adding files and directories to cache
for(let i = 0; i < input.length; i++) {
    let line = input[i].split(' ');
    if(line[0] == '$') {
        command(line);
    } else if (line[0] == 'dir') {
        cache[currentDirectory.join('/')].push(line[1]);
    } else {
        let size = line[0];
        let file = line[1];
        cache[currentDirectory.join('/')].push(size);
    }
}

// When we have our filesystem built from the input commands,
// we can loop over the cache and sum the size of each directory thats small enough.
let sum = 0;
currentDirectory = [];
for(let key in cache) {
    currentDirectory.push(key);
    let dirSize = sizeOfDirectory(cache[key]);
    if (dirSize <= 100000) {
        sum += dirSize;
    }
}
// console.log(cache);
console.log(sum);